Add Hosting TCO (#1476)
medic/cht-docs: content/en/hosting/costs/_index.md

---
title: "Hosting Costs"
linkTitle: "Hosting Costs"
weight: 2
aliases:

---

## Cost per CHT per Month

> In a production environment, the CHT costs $0.10 per user per month to run.  What exactly does this mean? Let's dive into some specifics which importantly cover what is include and excluded and how these costs can vary with each deployment.

Author: Yes - good point!  It is per active user, not per total users, so inactive users don't count. Explicit language added in [ca4ba6a](https://github.com/medic/cht-docs/pull/1476/commits/ca4ba6a89fcce416ffd2181259b21b80fbd2f1b4)

Be sure to read the [Accuracy section](#accuracy) so you understand what the costs on these page mean for your deployment.
{{% /pageinfo %}}

## How the number was calculated

Medic hosts a number of production CHT instances in Amazon [Elastic Kubernetes Service](https://aws.amazon.com/eks/) (EKS). By using [OpenCost](https://www.opencost.io/), Medic closely monitoring real world costs with actual end users. OpenCost allows us to separate one deployment's CPU, RAM and disk use from another's despite them all running in a multi-tenant EKS cluster.

By looking at dozens of instances and focusing on the highest usage ones, an accurate representation can be derived of a higher cost a self hosting partner.  

Here's some real costs for two week period for two instances that have been renamed to simply **A** and **B**:

| Per Month | A | B | 
| --- | --- | --- | --- |
| $/User | $0.05 | $0.10 |  |
| $ | $94 | $68 |  |
| Active Users (30 days) | 2000 | 661 |  |
| DB Size (Millions) | 7.5M | 10.7M |  |
| DB Growth (Thousands) | 123.2K | 248.0K |  |
| RAM | $16 | $22 |  |
| CPU | $6 | $8 |  |
| Storage | $53 | $24 |  |
| CPU max used (vCPUs) | 2.9 | 8 |  |
| RAM max (Gigs) | 20 | 15 |  |

> | Storage Max PVC (TBs) | 0.96 | 0.42 |  |

Author: This is great feedback, thanks!  I think we can reduce confusion by:
* remove **A** and **B** instances and going with just one
* Removing some rows from the table
* breaking the table up into two 
* making the math actually work by adding in over provisioning :sweat_smile: 
* adding an some more explanatory text

All done in 8623b8e06e299b01edbff486ed3952c63e591c1c

Deployment **A** has a lot of active users (`2000/mo`), but a slower database growth (`123.2K/mo`) as compared to deployment **B** which has fewer users (`661/mo`) but faster growth (`248K/mo`). These are the types of variations that can cause changes in costs mentioned in the next section.

### Accuracy

As our research was based on live instances, the costs are realistic.  However, since not all users will use the CHT in the exact same way and not all deployments have the same number of users and work flows, these numbers are still subject to some per-deployment fluctuation.

> Further, the costs in the table above only cover the variable costs.  That is, the costs incurred of actual usage with in an EKS cluster.  In a multi-tenant deployment where you have maybe three instances (Staging, User Acceptance Testing and Production), this can be critical to know what the three different deployments incurred costs are within your EKS cluster.  

Reviewer: ```suggestion
Further, the costs in the table above only cover the variable costs.  That is, the costs incurred of actual usage with in an EKS cluster.  In a multi-tenant deployment with multiple instances (for example: Staging, User Acceptance Testing and Production), this can be critical to know what the three different deployments incurred costs are within the EKS cluster.  
```

> Importantly, all production CHT 4.x deployments have a fixed cost to run the EKS cluster or generic Kubernetes if you're not on AWS. This is because all clusters contain servers that are always on and fully available to you. This cost is not yet included in this document and Medic hopes to include this cost over time.

Reviewer: ```suggestion
Importantly, all production CHT 4.x deployments have a fixed cost to run the EKS cluster or generic Kubernetes if not on AWS. This is because all clusters contain servers that are always on and fully available. This cost is not yet included in this document and Medic hopes to include this cost at a later date.
```

###  Production vs Development

A production deployment means the CHT is hosted either at a cloud provider or in a datacenter. Both provide high quality connectivity, power and cooling so the cost is higher than a development instance.  This ensures users do not have service interruptions and can continue to have trust the CHT will be up when they need to deliver care.

When analyzing the hosting total cost of ownership (TCO), only production instances were looked at.

It should be noted that a development environment can be no cost (or very low cost).  Assuming a developer already has a laptop, this is all that is needed to host a development instance. 

> Running an easy to set up Docker based development environment (instead of Kubernetes) on a laptop is an acceptable practice: there is no expectation of high uptime for development instances.

Reviewer: Consider linking to the app dev hosting docs.

Author: yup yup!  fixed in 9cd2749a

> ## Expenses behind the per user cost

Reviewer: ```suggestion
## What's included in the per user cost
```

> To be clear, this document is very much a hosting TCO and not a generic TCO document.  As such, below is a list of what's included in the hosting casts above and what is excluded.

Reviewer: ```suggestion
This document covers hosting cost and does not cover Total Cost of Ownership (TCO).  Below is a list of what's included in the hosting casts above and what is excluded.
```

### Included

Items that are included in the basic costs of hosting the CHT:

> * Servers - This is the most likely place where Medic can leverage their intimate knowledge from hosting many dozen production instances of the CHT.  This includes:

Reviewer: ```suggestion
* Servers - Given the large number of hosted CHT instances this is the area with the most data to use to build the calculator.  This includes:
```

* Built in overprovisioning to allow for bursts and a bit of growth
* Monitoring and Alerting - Since all software fails eventually, you need to be prepared to defend against this with aggressive monitoring and alerting.  The goal will be to fix the problem before any users notice.
* Ingress/Egress - This is the ability to send and receive data to the CHT.  Pricing from Datacenters and Cloud is readily available and "Servers" above can not exist without it.
* CHT Core updates and system maintenance - Both the CHT and underlying operating system will be upgraded on a regular basis.  This not only ensures there are no security vulnerabilities, but also ensures the deployments gains the benefit of new features and performance gains in the CHT.

### Excluded

There are many assumptions about what else is needed to run the CHT. While important, they are excluded:

* Kubernetes cluster - Current CHT production best practices include a Kubernetes cluster that will have a fixed monthly cost to operate, including the worker nodes which may be large in size (RAM & CPU).

> * Backups - Regular snapshots of the production data will need to be taken to ensure there is no data loss in case of catastrophic failure of server hardware.  This takes up a disk space which should be accounted for when budgeting to host the CHT.

Reviewer: This feels like one we should be able to include without too much effort - was the data no available?

> * Building a Datacenter (DC) - While some well funded MoH's may have the budget and time to build a complete data center, this is out of the scope of Medic's core competency.  Therefore it is assumed an existing DC will be used or cloud hosting will be used.  In the case where a MoH wants to build a DC, a competent 3rd party should be selected

Reviewer: ```suggestion
* Building a Datacenter (DC) - While some well funded MoH's may have the budget and time to build a complete data center, this is out of the scope for this calculation.  Therefore it is assumed an existing DC will be used or cloud hosting will be used.  In the case where a MoH wants to build a DC, a competent 3rd party should be selected
```

* Training of Systems Administrators/IT - System administrator IT systems that have not used Kubernetes or have not hosted the CHT, will need to be trained on how to do both.

> * Training of Trainers (ToT) and user Training -

Reviewer: ```suggestion
* Training of Trainers (ToT) and user Training
```

* Data Warehousing & Dashboards - While close to "Servers" above, not all deployments have a data warehouse (CHT Sync + Postgres) and dashboards (Superset/Klipfolio) at launch.  These are easy to add on at a later date and cost can be estimated at that time.
* Upfront Purchase of Hardware - It is assumed that a deployment will either be using cloud based solutions or using managed bare metal, so these costs don't apply.
* App Development - Each deployment needs to have the default CHT app customized for each of the required workflows.
* Smartphones - Device purchase, setup, and distribution for users to use the CHT.
* Analog -> Digital Workflow conversion - The process of documenting paper processes in places that the CHT will replace.
* SMS - Some projects need the ability to send SMS to users from the CHT.
* Interoperability - While the CHT supports this out of the box - development work is needed to ensure it works with the specific 3rd party systems.
* Mobile Device Management - optional.
* Cellphone telecom services - Users actually uploading/downloading data

> * Transfer of hosting from Medic hosted to MoH Self Hosted - Many person hours needed to help with transfer of data to new MoH based hosting.

Reviewer: Recommend this is removed as it's a one off. Or possibly changed to talk about server set up rather than calling out Medic specifically. eg:

```suggestion
* Setup and/or transfer of hosting provider
```
